### PR TITLE
Fix props factory hover and go to definition behavior

### PR DESCRIPTION
**Pull request by vilhena-stripe**
### Motivation
Hover and go to definition for `factory` in `prop` would send user to the beginning of the `prop` line. Fixes https://github.com/sorbet/sorbet/issues/2188

### Test plan
See included automated tests.

### CODE REVIEW

File: rewriter/Prop.cc [hunk 1]
```diff
@@ -232,6 +232,8 @@ optional<NodesAndPropInfo> processProp(core::MutableContext ctx, ast::Send *send
                 }
             }
         }
+
+        ret.nodes.emplace_back(ast::MK::Unsafe(send->loc, rules->deepCopy()));
```
Comment (jvilk-stripe) [inline rewriter/Prop.cc]: Note: We tried using keepForIde, but the verifier complained that the keepForIde node lacked loc information.

We then added in an exclusion for keepForIde in that pass, but then it complained that `Constant(loc, core::Symbols::Sorbet_Private_Static())`, which keepForIde creates, lacks loc information.

It seemed like we were fighting an uphill battle, and that T.unsafe might be the safest option.

**Comment by jvilk-stripe** [comment]
This change might not be completely correct; we've discovered that some (not sure if all?) lambda arguments to `props` expect to run with `self` set to the owning object. And since the stuff we're adding to the AST is subject to typechecking, it's causing new errors to pop up.

I have a link to a typecheck on Stripe's internal code I can share if you want to see. I have a feeling this issue would be present if we used `keepForIde` also.

We could explicitly model `render_field_proc` (e.g., set its receiver properly), which seems to be the prime problematic keyword argument.

**Comment by elliottt** [comment]
Just checking in, is this PR still active?

**Comment by jvilk-stripe** [comment]
@elliottt I think we'd like to take it across the finish line, but we're not sure what the correct approach is.

We need to keep more stuff in the AST so that the IDE can respond to requests concerning it, but that means Sorbet typechecks it. Some `prop` properties expect `self` to be bound to the instance type, and fail typechecking in pay-server with this change. Do we whitelist specific props that we understand and ignore the rest? Is our approach acceptable, or should we make keepForIde work for rewriter?

**Comment by DarkDimius** [comment]
@jvilk-stripe and @vilhena-stripe could you please take an action to extract tests from Stripe internal repo that would have failed and add them to this repo?

**Comment by jez** [comment]
This is already fixed, via #2931. I've added a test to ensure that we don't regress in #2996.

**Comment by jvilk-stripe** [comment]
Thanks Jake! I'll have to take a look at your solution to see how you did this. :)

**Pull request closed without merging**